### ISSUE TITLE
Destructuring, defaults and a type confuse the linter

### ISSUE THREAD

**Issue by djMax** [yannickcr/eslint-plugin-react#1197]
This is a messy one...

```
type Props = {
  history: {
    push: Function,
  },
  match?: {
    params?: {
      date?: string
    },
  },
};

export const Daily = ({ history, match: {
  params: {
    date = moment().toISOString(),
  } = {},
} = {} }: Props) => (
      <div>
        <div style={{ textAlign: 'right', paddingRight: '25px' }}>
          Show <DatePicker
            selected={moment(date)}
            className="datetime"
            onChange={d => history.push(`./${d.toISOString()}`)}
          />
        </div>
        <WithData url="/payments/daily" body={{ date: moment(date).toISOString() }}>
          <Flashcards date={moment(date)} />
        </WithData>
      </div>
);
```

Yields these seemingly spurious lint errors:
```
  11:11  error  'history.push' PropType is defined but prop is never used                           react/no-unused-prop-types
  13:3   error  propType "match" is not required, but has no corresponding defaultProp declaration  react/require-default-props
  15:14  error  'match.params.date' PropType is defined but prop is never used                      react/no-unused-prop-types
```

**Comment by ljharb** [comment]
The no-unused-prop-types ones are spurious, and fixed in latest master.

The require-default-props is correct, because argument defaults aren't sufficient - the rule wants `.defaultProps` defined.
